File: 337.house-robber-iii.py
#
# @lc app=leetcode id=337 lang=python3
#
# [337] House Robber III
#

# Time:  O(n)
# Space: O(h)


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def rob(self, root: TreeNode) -> int:
        def helper(node: TreeNode):  # -> (node is robbed, node is not robbed)
            if not node:
                return (0, 0)
            left, right = helper(node.left), helper(node.right)
            return (node.val + left[1] + right[1], max(left) + max(right))

        return max(helper(root))


# @lc code=end

File: 337.house-robber-iii.js
/*
 * @lc app=leetcode id=337 lang=javascript
 *
 * [337] House Robber III
 */

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start
/**
 * @param {TreeNode} root
 * @return {number}
 */
var rob = function (root) {
  const helper = (node) => {
    if (!node) {
      return [0, 0];
    }
    const left = helper(node.left);
    const right = helper(node.right);
    return [
      node.val + left[1] + right[1],
      Math.max(...left) + Math.max(...right),
    ];
  };
  return Math.max(...helper(root));
};
// @lc code=end
